contrib/babelfishpg_tsql/src/pl_funcs-2.c

```diff
@@ -351,58 +352,84 @@ coalesce_typmod_hook_impl(const CoalesceExpr *cexpr)
+ */
 void
-check_restricted_stored_procedure(Oid proc_id)
+check_restricted_object(Oid object_id, ObjectType object_type)
 {
-	HeapTuple		proctup;
-	Form_pg_proc		procform;
-	const char		*procname;
-	Oid		schema_oid = InvalidOid;
-	Oid		dbo_oid = InvalidOid;
-	bool		is_restricted = false;
+	HeapTuple	tuple;
+	char		*objname;
+	char		*schema_name;
+	Oid			schema_oid;
 
-	/*
-	 * List of procedure names that are not allowed to be dropped. These procedures 
-	 * are considered essential or restricted due to security or operational reasons.
-	 */
-	static const char	*restricted_procedures[] = {
-		"xp_qv",
-		"xp_instance_regread",
-		"sp_addlinkedsrvlogin",
-		"sp_droplinkedsrvlogin",
-		"sp_dropserver",
-		"sp_enum_oledb_providers",
-		"sp_testlinkedserver"
-	};
-	static const int	RESTRICTED_PROCEDURES_COUNT = sizeof(restricted_procedures) / sizeof(restricted_procedures[0]);
-
-	proctup = SearchSysCache1(PROCOID, ObjectIdGetDatum(proc_id));
-
-	procform = (Form_pg_proc) GETSTRUCT(proctup);
-	procname = pstrdup(NameStr(procform->proname));
-	schema_oid = procform->pronamespace;
-	dbo_oid = get_namespace_oid("master_dbo", true);
-
-	if (OidIsValid(schema_oid) && OidIsValid(dbo_oid) && schema_oid == dbo_oid)
+	/* Get object information */
+	if (object_type == OBJECT_PROCEDURE || object_type == OBJECT_FUNCTION)
+		tuple = SearchSysCache1(PROCOID, ObjectIdGetDatum(object_id));
+	else
+		tuple = SearchSysCache1(RELOID, ObjectIdGetDatum(object_id));
+
+	if (!HeapTupleIsValid(tuple))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("object with OID %u does not exist", object_id)));
+
+	/* Get object name and schema */
+	if (object_type == OBJECT_PROCEDURE || object_type == OBJECT_FUNCTION)
 	{
-		/* Check if the procedure name is in the restricted list */
-		for (int i = 0; i < RESTRICTED_PROCEDURES_COUNT; i++)
-		{
-			if (pg_strcasecmp(procname, restricted_procedures[i]) == 0)
-			{
-				is_restricted = true;
-				break;
-			}
-		}
+		Form_pg_proc procform = (Form_pg_proc) GETSTRUCT(tuple);
+		objname = pstrdup(NameStr(procform->proname));
+		schema_oid = procform->pronamespace;
+	}
+	else
+	{
+		Form_pg_class classform = (Form_pg_class) GETSTRUCT(tuple);
+		objname = pstrdup(NameStr(classform->relname));
+		schema_oid = classform->relnamespace;
+	}
+
+	/* Get schema name */
+	if (OidIsValid(schema_oid))
+		schema_name = get_namespace_name(schema_oid);
+	else
+	{
+		ReleaseSysCache(tuple);
+		return;
 	}
 
-	ReleaseSysCache(proctup);
-	if (is_restricted)
+	/* Check if object is restricted and error out if it is */
+	if (is_object_restricted(objname, schema_name))
 	{
+		if (schema_name)
+			pfree(schema_name);
 		ereport(ERROR,
 				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
-				 errmsg("must be owner of procedure %s", procname)));
+				 errmsg("must be owner of %s %s",
+				 		 object_type == OBJECT_PROCEDURE ? "procedure" :
+						 object_type == OBJECT_VIEW ? "view" :
+						 "function",
+						 objname)));
```

Comment:
A suggestion: This error message is confusing since the owner of these system objects is 'sysadmin' and we won't allow sysadmin(or its member) also to drop these objects either.  "must be owner" is a generic message that comes from PG, lets make it Babelfish specific.
Better error message would be: `Babelfish-created system view/procedure/function cannot be dropped outside of a Babelfish session` similar to what we throw for BBF roles.